### PR TITLE
platform-schema: CI hardening, anti-drift hardening, owner-model docs, and trigger coverage

### LINKED ISSUE #1337: ci(platform-schema): expand conformance workflow path triggers to cover adapter source code
## Why

The anti-drift conformance tests validate that every `source` code-ref in platform schema files (e.g. `crates/openab-gateway/src/adapters/line.rs#dispatch_line_reply`) still points at a real file + symbol in the tree. This is the **core value proposition** of the platform schema system — docs that break the build when they go stale.

However, the CI workflow currently only triggers when schema files themselves change. If someone renames `dispatch_line_reply` to `send_line_message` in an adapter PR, the conformance workflow **never runs**. The dead code-ref silently lands on main, and the next unrelated schema PR fails with a confusing "symbol not found" error (action at a distance).

**Without this fix, the anti-drift mechanism is blind to the most common drift scenario: code refactoring.**

## Benefits

- **Anti-drift actually works end-to-end** — symbol renames/deletes in adapter code are caught at PR time, not weeks later
- **No more "test poisoning"** — unrelated schema PRs won't randomly fail due to accumulated dead refs from other PRs
- **Confidence in schema accuracy** — if CI is green, the code-refs are valid (currently green CI only means "nobody touched schema files")
- **Negligible cost** — the conformance job runs in ~21 seconds (only serde + toml deps), far cheaper than a flaky test

## Proposed Fix

Expand path triggers in `.github/workflows/platform-schema-conformance.yml`:

```yaml
paths:
  - "docs/platforms/schema/**"
  - "docs/platforms/_template.toml"
  - "crates/platform-schema/**"
  - "crates/openab-gateway/src/**"   # code-refs point here
  - "crates/openab-core/src/**"      # code-refs point here
  - "src/main.rs"                    # cron_adapters ref
```

Or remove the path filter entirely — the job is fast enough to run on all PRs.

Additionally, add a weekly scheduled run as safety net:
```yaml
schedule:
  - cron: "0 6 * * 1"
```

## Context

Identified during group review of #1295 (8/8 reviewers flagged this independently). Ref: #1294.

### LINKED ISSUE #1338: ci(platform-schema): add --locked, Rust cache, checkout@v6, and concurrency group
## Why

The `platform-schema-conformance.yml` workflow works but has several rough edges that affect reliability, speed, and consistency with the rest of the repo's CI infrastructure.

1. **No `--locked`** — `cargo test` without `--locked` could silently update `Cargo.lock` if upstream publishes compatible versions, making builds non-reproducible
2. **No build cache** — every run recompiles `serde` + `toml` from scratch (~30-60s wasted)
3. **`actions/checkout@v4`** — the rest of the repo uses `@v6`; inconsistency causes noise during repo-wide action bumps and misses security improvements shipped in v6
4. **No concurrency group** — rapid pushes (e.g. contributor fixing schema errors in sequence) queue duplicate runs, wasting CI minutes

## Benefits

- **Reproducible builds** — `--locked` guarantees the same dependency versions every run
- **30-60s faster CI** — Rust cache eliminates redundant compilation
- **Consistent security posture** — checkout@v6 `persist-credentials` improvement
- **Cleaner CI tab** — concurrency group auto-cancels superseded runs, contributor sees only the latest result
- **Better contributor DX** — faster feedback loop on schema changes

## Proposed Changes

```yaml
name: platform-schema conformance

concurrency:
  group: platform-schema-${{ github.ref }}
  cancel-in-progress: true

jobs:
  conformance:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v6
      - uses: dtolnay/rust-toolchain@stable
      - uses: Swatinem/rust-cache@v2
        with:
          workspaces: "crates/platform-schema -> target"
      - run: cargo test --locked --manifest-path crates/platform-schema/Cargo.toml
```

## Context

Identified during group review of #1295. Ref: #1294.

### LINKED ISSUE #1339: docs(platforms): assign page owners and establish maintenance guidelines
## Why

The platform schema knowledge base (`docs/platforms/schema/*.toml`) has 8 platform files with rich capability/feature documentation, but:

1. **No assigned owners** — when a platform's API changes or our adapter evolves, nobody is specifically responsible for updating the schema
2. **No PR checklist** — contributors modifying adapter code aren't reminded to update the corresponding schema, so drift accumulates silently
3. **Unclear dual-docs relationship** — new contributors see both `docs/line.md` and `docs/platforms/schema/line.toml` and don't know which to update or why both exist

Without explicit ownership and process, the schema files will become "passing CI but factually stale" documentation — exactly the problem they were created to solve.

## Benefits

- **Clear accountability** — each platform has someone who ensures facts stay current
- **Drift prevention at PR time** — checklist reminds contributors to update schema when changing adapter behavior
- **Zero onboarding confusion** — new maintainers immediately understand: TOML = machine facts, Markdown = human setup guide
- **Sustainable long-term** — without this, 2800 lines of TOML will silently rot within 6 months

### 1. Assign platform page owners

| Platform | Owner |
|----------|-------|
| line | TBD |
| slack | TBD |
| telegram | TBD |
| discord | TBD |
| feishu | TBD |
| wecom | TBD |
| googlechat | TBD |
| teams | TBD |

### 2. Add to PR template / checklist

When the PR touches `crates/openab-gateway/src/adapters/*.rs`:
- [ ] Updated `docs/platforms/schema/<platform>.toml` if feature status or capabilities changed

### 3. Document dual-docs architecture in README

Already partially done in #1295 (architecture diagram + "How to update" section). Verify the explanation is clear and complete.

## Context

Identified during group review of #1295. Per #1294: "Each per-platform page is owned by that platform's maintainer."

### LINKED ISSUE #1340: fix(platform-schema): harden check_code_ref — reject empty symbols and path traversal
## Why

`check_code_ref` in `crates/platform-schema/tests/conformance.rs` is the anti-drift engine — it validates that every `source` code-ref in platform schema files still points at a real file and symbol. However, it has edge cases that reduce trust in the validation:

1. **Empty symbol bypass** — `"file.rs#"` (trailing `#` with no symbol) parses to `symbol = Some("")`. Since `text.contains("")` always returns `true` in Rust, the check silently passes. A typo or copy-paste error could create a ref that looks validated but isn't.

2. **Path traversal** — `root.join(r.file)` accepts `../` components. A source like `"../../etc/passwd"` would pass `is_file()`. While PR review would likely catch this, defense-in-depth should reject it programmatically.

3. **Inconsistent URL handling** — `feature_sources_exist_in_tree` rejects non-code-ref sources with a confusing "expected a file ref" error. `quirk_code_sources_exist_in_tree` correctly uses `is_code_ref()` to skip URLs. If a contributor adds a URL to a feature source, they get a misleading error.

## Benefits

- **Higher confidence in anti-drift** — no silent bypass routes exist; if CI passes, refs are genuinely validated
- **Defense-in-depth** — even if a bad ref slips through review, CI catches it
- **Better contributor DX** — clear error messages for each failure mode (not a raw Rust panic)
- **Consistent behavior** — feature and quirk source validation follow the same logic

### 1. Reject empty symbol

```rust
if let Some(sym) = r.symbol {
    if sym.is_empty() {
        return Err("empty symbol after '#' is not allowed".to_string());
    }
    let text = fs::read_to_string(&path).map_err(|e| e.to_string())?;
    if !text.contains(sym) {
        return Err(format!("symbol {sym:?} not found in {:?}", r.file));
    }
}
```

### 2. Prevent path traversal

```rust
let resolved = root.join(r.file).canonicalize().map_err(|e| e.to_string())?;
if !resolved.starts_with(root) {
    return Err(format!("source path {:?} escapes repo root", r.file));
}
```

### 3. Align URL handling

In `feature_sources_exist_in_tree`, add before the error push:
```rust
if !is_code_ref(src) {
    continue;  // skip URLs, same as quirk test
}
```

## Context

Identified during group review of #1295 (B1, B4, B5 flagged independently). Ref: #1294.

### PR DESCRIPTION
**Pull request by luffy-aiagent**
## What problem does this solve?

Closes #1337
Closes #1338
Closes #1339
Closes #1340

Discord Discussion URL: https://discord.com/channels/1491295327620169908/1491365158868619404/1532377620241449040

All four issues were identified during group review of #1295 (the platform-schema knowledge base PR) and are bundled here since they're small, related follow-ups to that same subsystem.

### Goal

1. **#1337** — Expand `platform-schema-conformance.yml`'s trigger paths to also cover the adapter/core source trees that schema code-refs point into, plus a weekly scheduled run as a backstop.
2. **#1338** — Harden the same workflow: `actions/checkout@v6` (was `@v4`, inconsistent with rest of repo), add `Swatinem/rust-cache`, add a `concurrency` group.
3. **#1340** — Harden `check_code_ref` in `crates/platform-schema/tests/conformance.rs`: reject empty `#symbol`, reject path traversal outside repo root.
4. **#1339** — Replace the proposed fixed per-platform-owner model with open contribution guidance + a documented known limitation (see rationale below and issue comment).

### Non-goals

- #1340's third sub-point (aligning `feature_sources_exist_in_tree` to skip URL sources like the quirk check does) is intentionally **not** included — see "Alternatives Considered".
- No named per-platform owners are assigned (#1339) — see issue comment for rationale.

### Accepted Residual Risks

- [x] `feature_sources_exist_in_tree` still hard-errors on a URL source instead of skipping it (unlike quirks). Verified no current schema file needs this (136 feature `source` entries checked, none are URLs). Documented as an explicit decision in `docs/platforms/README.md`, revisit if a feature genuinely needs a URL source.
- [x] Semantic drift (a code-ref stays valid but the `note`/`status` it documents goes stale) is not detectable by CI — only syntactic drift (renamed/removed symbols) is, and only for the file trees now covered by #1337's trigger paths. Documented as a known limitation; no dedicated owner audits for it.

### Acceptance Criteria

- [x] `platform-schema-conformance.yml` triggers on changes to `crates/openab-gateway/src/**`, `crates/openab-core/src/**`, and `src/main.rs` (all current schema code-ref targets, verified by extracting every `.rs` path cited across all 8 schema files), plus a weekly `schedule` backstop paired with `workflow_dispatch` for manual testing (matches the other 10 schedule-triggered workflows in this repo)
- [x] `platform-schema-conformance.yml` uses `checkout@v6`, has a Rust build cache, and a concurrency group keyed by both `github.event_name` and `github.ref` (a `push`-to-main run and the weekly `schedule` run both resolve `github.ref` to `refs/heads/main`; without `event_name` in the key they'd share a group and `cancel-in-progress` could cancel one in favor of the other)
- [x] `check_code_ref` rejects `"file.rs#"` (empty symbol) with a clear error
- [x] `check_code_ref` rejects sources that resolve outside the repo root, including both `../`-style relative traversal and a bare absolute path (`Path::join` replaces the base entirely for an absolute joined path — a distinct code path from relative traversal)
- [x] Three new tests cover empty-symbol, relative-traversal, and absolute-path cases
- [x] `CONTRIBUTING.md` invites any contributor to update platform schema TOMLs directly, without needing a dedicated owner
- [x] `docs/platforms/README.md` documents both the URL-source decision and the semantic-drift known limitation

### Follow-ups

- If semantic-drift incidents actually occur in practice, open a new issue at that point to discuss the solution in light of the concrete situation, rather than presupposing per-platform ownership is the right fix.

## At a Glance

```
CI workflow + Rust test hardening + docs — no runtime/architecture impact
```

## Prior Art & Industry Research

Not applicable — CI config, test hardening, and docs changes only; no runtime/architectural/delivery behavior affected.

## Proposed Solution

**#1337** (`.github/workflows/platform-schema-conformance.yml`):
- Added `crates/openab-gateway/src/**`, `crates/openab-core/src/**`, `src/main.rs` to both `pull_request` and `push` path triggers — confirmed these cover every `.rs` file currently cited by a schema code-ref (extracted all `source` refs across the 8 platform files)
- Added `schedule: cron: "0 6 * * 1"` (weekly) as a backstop in case a PR touching the source tree somehow bypasses the path triggers, paired with `workflow_dispatch:` so it can also be triggered manually (matches every other schedule-triggered workflow in this repo)

**#1338** (same file):
- `actions/checkout@v4` → `@v6` (matches the rest of the repo's workflows)
- Add `Swatinem/rust-cache` (same pinned SHA used elsewhere in the repo) scoped to `crates/platform-schema -> target`
- Add `concurrency: { group: platform-schema-${{ github.event_name }}-${{ github.ref }}, cancel-in-progress: true }` — `event_name` is included because `push` and `schedule` runs both resolve `github.ref` to `refs/heads/main`; without it they'd share a group and could cancel each other
- `--locked` was already present from a prior change; left untouched

**#1340** (`crates/platform-schema/tests/conformance.rs`):
- `check_code_ref` now canonicalizes both `root` and the resolved `path` and checks `starts_with` before proceeding, rejecting any source that escapes the repo root, whether via `../..` relative traversal or a bare absolute path (`Path::join` replaces the base entirely when the joined path is absolute — a distinct code path from relative traversal)
- `check_code_ref` now rejects an empty `#symbol` explicitly instead of silently passing (`text.contains("")` is always `true` in Rust)
- Three new tests: `check_code_ref_rejects_empty_symbol`, `check_code_ref_rejects_path_traversal`, `check_code_ref_rejects_absolute_path`

**#1339** (`CONTRIBUTING.md`, `docs/platforms/README.md`):
- `CONTRIBUTING.md`: added a sentence inviting any contributor to update the schema TOML directly when they notice a platform API change or quirk, rather than waiting on a dedicated owner
- `docs/platforms/README.md`: added a "Known limitation" note under Conformance about semantic vs syntactic drift

## Why this approach?

**#1337/#1338/#1340** are straightforward hardening matching patterns already used elsewhere in this repo's CI (see `ci.yml`, `docker-smoke-test.yml`) and Rust code (canonicalize-and-check is the standard path-traversal guard). #1337's path list was derived by actually grepping every code-ref in the 8 schema files rather than guessing, to make sure the trigger coverage is complete as of today.

**#1339**: discussed with @sky092879 — two of the issue's three asks were already satisfied by existing docs (`CONTRIBUTING.md:202`, README's "Architecture: TOML vs Markdown" section). For the remaining ask, a formal per-platform ownership table was judged to add coordination overhead disproportionate to the current team size, with no clear enforcement mechanism. The underlying concern (stale docs) splits into syntactic drift — now handled by #1337's expanded triggers — and semantic drift, which has no automated solution and is now an explicit, revisit-if-needed limitation instead of an unaddressed gap.

## Alternatives Considered

- **Relax `feature_sources_exist_in_tree` to skip URL sources** (issue #1340's third sub-point): considered, rejected for now. No current feature source is a URL (checked all 8 schema files, 136 entries), so this would be a preemptive change with no current benefit, and could let feature sources silently drift to unverifiable doc links instead of code. Documented as an explicit decision in `docs/platforms/README.md`; revisit if a feature genuinely needs a URL source.
- **Non-blocking PR-comment bot** reminding reviewers to check schema accuracy when adapter files change (for #1339's semantic-drift gap): considered, deferred in favor of the lighter-weight CONTRIBUTING.md guidance + #1337's CI trigger fix.
- **Assign named per-platform owners** (#1339's original proposal): rejected for now — see rationale above.
- **Remove the path filter entirely** (#1337's alternative, run conformance on every PR): rejected in favor of an explicit path list — the job is cheap either way, but an explicit list documents *why* it runs, and the weekly schedule already covers the "something slipped through" case.

## Validation

Rust changes:
- [ ] `cargo check` passes — not verified in this environment (no Rust toolchain available); please verify in CI
- [ ] `cargo test` passes (including new tests) — not verified in this environment; please verify in CI
- [ ] `cargo clippy` clean — not verified in this environment; please verify in CI

CI/workflow changes:
- [x] Workflow syntax reviewed by hand (no local `actions`/YAML linter available in this environment)

Docs-only changes:
- [x] Links are valid
- [x] Renders correctly in GitHub preview

All PRs:
- [x] Manual testing — read through the diff and rendered Markdown; traced `check_code_ref`'s new logic against the empty-symbol, relative-traversal, and absolute-path cases by hand, and cross-checked #1337's new trigger paths against every code-ref actually cited in the 8 schema files, since `cargo test` could not be run locally.
- [x] Independently reviewed by 4 focused passes (Rust correctness, CI/workflow correctness, docs consistency, security) before this revision — caught and fixed the `concurrency` group collision between `push` and `schedule`, the missing `workflow_dispatch`, and the missing absolute-path test case.

### CODE REVIEW

**Comment by feiyun968-agent** [comment]
## PR Review — 司馬

### Doc Issues

**🟢 `docs/platforms/README.md` — Decision block**
新增的 `> **Decision:**` blockquote 清楚說明 `feature_sources_exist_in_tree` 故意拒絕 URL sources，並附上 issue #1340 供追蹤。內容正確，無問題。

**🟢 `docs/platforms/README.md` — Known limitation**
Semantic drift 的 known limitation 描述準確，邊界清楚：CI 只能抓 syntactic drift，semantic drift 靠 PR review。無問題。

**🟡 `CONTRIBUTING.md` — 新增段落位置略重疊**
新段落插在兩段既有內容之間：
- 上文：「修改 adapter 時，更新對應的 .toml」
- 新增：「發現 API 變了或有新 quirk，直接更新 .toml，不需等 owner」
- 下文：「See docs/platforms/README.md for...」

兩段語氣與對象略有重疊，可考慮合併為一段或加小標，讓意圖更清晰。非 blocker。

**🟢 Cross-doc consistency**
`CONTRIBUTING.md` 與 `docs/platforms/README.md` 對 semantic drift 的描述互相對應，無矛盾。

### Summary

| 項目 | 結果 |
|------|------|
| Doc 內容正確性 | ✅ |
| Doc 與 code 一致性 | ✅ |
| Doc 彼此一致 | ✅ |
| 建議 | 🟡 `CONTRIBUTING.md` 兩段可考慮合併，非 blocker |

**整體：無 blocking doc issue。**

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -201,6 +201,8 @@ gh run view <run_id> --repo openabdev/openab --json conclusion -q .conclusion
 
 When modifying a platform adapter (`crates/openab-gateway/src/adapters/*.rs`), check whether the change affects the platform's documented capabilities or feature status. If it does, update the corresponding `docs/platforms/schema/<platform>.toml`.
 
+If you notice a platform's official API has changed (a new webhook event type, a deprecated field, etc.) or spot a platform-specific quirk while working on something else, feel free to update the corresponding `docs/platforms/schema/<platform>.toml` (`[capability.*]` or `[[quirks]]`) directly in your PR — no need to wait for a dedicated owner to notice.
```
Comment (thepagent) [inline CONTRIBUTING.md]: we should make this assertive

**Review by thepagent — APPROVED** [review]
